### ISSUE TITLE
pc.string.startsWith could be dramatically improved

### ISSUE THREAD

**Issue by Maksims** [playcanvas/engine#222]
Using `String.indexOf` for `String.startsWith` is overkill.  
This method is used in some places with loops and potentially by users in update game loop.

Method with loop and early break performs about 350+ times faster, check out jsperf here https://jsperf.com/string-startswith/21

Consider replacing startsWith with native (ECMA6) implementation (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith) and polyfill:
```javascript
// String.startsWith
if (! String.prototype.startsWith) {
    Object.defineProperty(String.prototype, 'startsWith', {
        enumerable: false,
        configurable: false,
        writable: false,
        value: function(str) {
            var that = this;
            for(var i = 0, ceil = str.length; i < ceil; i++)
                if(that[i] !== str[i]) return false;
            return true;
        }
    });
}
```

**Comment by daredevildave** [comment]
Sure. I think we should internalize this API. We're not making a general purpose JS lib, I don't want to be supporting string methods, etc.

**Comment by willeastcott** [comment]
I wouldn't even internalize it. Like @Maksims says, startsWith already works in Chrome and FF. We simply polyfill for other browsers. We should remove endsWith too. Also note that pc.string.toBool is not used anywhere in the engine. And pc.string.format is pretty worthless. Just use the string + operator, fer cryin' out loud. :)

**Comment by daredevildave** [comment]
pc.string.format is super-useful...! Maybe not in engine though.

Dave Evans
CTO, playcanvas.com
@playcanvas <http://twitter.com/playcanvas>
@daredevildave <http://twitter.com/daredevildave>
---
It's not just me who thinks we're great:
Technical Innovation Award Nomination
<http://www.developawards.com/finalists2>
Top 100 Game Tech Company
<http://www.develop-online.net/news/develop-100-27-uk-games-hardware-and-tools-named-in-tech-list/0194913>
Featured by Apple at WWDC
<https://developer.apple.com/videos/wwdc/2014/#509>

**Comment by willeastcott** [comment]
I think 'super-useful' is maybe overstating it. Let's look at the only usage not related to printing errors/warnings in the engine:

                var path = this.name;
                var format = "{0}/{1}";

                while (parent && parent._parent) {
                    path = pc.string.format(format, parent.name, path);
                    parent = parent._parent;
                }

How is that better than:

                var path = this.name;
                while (parent && parent._parent) {
                    path = parent.name + '/' +  path;
                    parent = parent._parent;
                }

**Comment by Maksims** [comment]
Implemented in #379
